### ISSUE TITLE
Traffic from the pod's local node

### ISSUE THREAD

**Issue by zerkms** [cloudnativelabs/kube-router#803]
I've found that allowing traffic from `match src-type LOCAL` is added after https://github.com/cloudnativelabs/kube-router/issues/87

That issue refers to the https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#networkpolicyspec-v1-networking-k8s-io where it states

> List of ingress rules to be applied to the selected pods. Traffic is allowed to a pod if there are no NetworkPolicies selecting the pod (and cluster policy otherwise allows the traffic), OR if the traffic source is the pod's local node, OR if the traffic matches at least one ingress rule across all of the NetworkPolicy objects whose podSelector matches the pod. If this field is empty then this NetworkPolicy does not allow any traffic (and serves solely to ensure that the pods it selects are isolated by default)

I want to emphasise this very part: `OR if the traffic source is the pod's local node`

I think that part was misinterpreted: I believe it means that if the traffic goes from the node IP address (eg originates from the process that runs on the host machine itself and has socket bound to one of the physical network interfaces). Not that it means `match src-type LOCAL` which effectively means **every process including containers**.

In other words, I think that it should allow packets that originate from the host local IP interfaces, not from literally all of them, since it renders network policies on that node for the pods on that node useless.

I think that requirement exists there so that kubelet could run readiness/liveness probes. For that purpose just whitelisting the node IP is enough.

Please let's discuss it.

**Comment by andrewsykim** [comment]
Hi @zerkms 

Looks like the original PR did use node IP https://github.com/cloudnativelabs/kube-router/pull/195, but a subsequent change updated it to `src-type LOCAL`. This is the commit but there's no justification for the change as far as I can tell https://github.com/cloudnativelabs/kube-router/commit/84741b6a5eaa9a29846b9b7fdf1e56e07faba4c2

**Comment by zerkms** [comment]
@murali-reddy could you please share some insights on why that change has been done?

**Comment by andrewsykim** [comment]
@zerkms @murali-reddy is on leave at the moment :) 

Were you able to confirm that any pod on your node is able to access the pod IP or was this an observation based on looking through the code?

**Comment by zerkms** [comment]
@andrewsykim I was able to confirm it:

1. I have disabled all ingress traffic for the app that is behind an `ingress-nginx`
1. I expected the app to be instantly unavailable (with http 502 or something similar)
1. I have found instead that just half of the requests to the backend app is rejected - those were from the ingress-nginx instance from a different node, while an ingress-nginx instance on the same node successfully connected (while I expected it should not have)

Then I checked iptables to find that it has that rather permissive rule.

**Comment by andrewsykim** [comment]
Can you share the iptable rules you have there? (easier for me than digging through the code)

**Comment by zerkms** [comment]
```
Chain KUBE-NWPLCY-KYPR6GDNRTNCK732 (3 references)
    pkts      bytes target     prot opt in     out     source               destination     

Chain KUBE-NWPLCY-X4CHFVVG7J2ROIVY (1 references)
    pkts      bytes target     prot opt in     out     source               destination         
       1       60 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* rule to ACCEPT traffic from source pods to dest pods selected by policy name app-from-ingress namespace <redacted> */ match-set KUBE-SRC-QFEMIHEFQPMVWYV3 src match-set KUBE-DST-LF3JBBI4OR2LIKWS dst tcp dpt:8080

Chain KUBE-POD-FW-APC7AWYDIVNYIYM3 (3 references)
    pkts      bytes target     prot opt in     out     source               destination         
      46    17570 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* rule for stateful firewall for pod */ ctstate RELATED,ESTABLISHED
       4      240 ACCEPT     all  --  *      *       0.0.0.0/0            10.71.6.179          /* rule to permit the traffic traffic to pods when source is the pod's local node */ ADDRTYPE match src-type LOCAL
       1       60 KUBE-NWPLCY-KYPR6GDNRTNCK732  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* run through nw policy default-ingress-deny */
       1       60 KUBE-NWPLCY-X4CHFVVG7J2ROIVY  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* run through nw policy app-from-ingress */
       0        0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* default rule to REJECT traffic destined for POD name:<redacted> namespace: <redacted> */ reject-with icmp-port-unreachable
```

And ipsets contain corresponding ingress and app IPs (not sure posting them would help).

**Comment by zerkms** [comment]
You know what, it looks like it's my terrible mistake: those requests from the ingress have been allowed through `ctstate RELATED,ESTABLISHED` not through `ADDRTYPE match src-type LOCAL`.

My apologise, I'm closing it for now.

**Comment by andrewsykim** [comment]
@zerkms no problem, appreciate the intent :)
